File: obsidian-plugin/src/providers/anthropic.ts
import Anthropic from "@anthropic-ai/sdk";
import type { LlmProvider } from "./base";
import { HANDWRITTEN_NOTES_PROMPT } from "../prompt";

const SYSTEM_PROMPT = HANDWRITTEN_NOTES_PROMPT;

export class AnthropicProvider implements LlmProvider {
  private client: Anthropic;
  private model: string;

  constructor(apiKey: string, model: string) {
    this.client = new Anthropic({ apiKey, dangerouslyAllowBrowser: true });
    this.model = model;
  }

  async ocr(imageDataUrls: string[], outputFormat: "markdown" | "text"): Promise<string> {
    const content: Anthropic.MessageParam["content"] = [];

    imageDataUrls.forEach((dataUrl, i) => {
      const match = dataUrl.match(/^data:(image\/\w+);base64,(.+)$/);
      if (!match) return;
      const [, rawMediaType, b64] = match;
      const mediaType = rawMediaType as "image/png" | "image/jpeg" | "image/webp" | "image/gif";

      if (imageDataUrls.length > 1) {
        content.push({ type: "text", text: `[Page ${i + 1}]` });
      }
      content.push({
        type: "image",
        source: { type: "base64", media_type: mediaType, data: b64 },
      });
    });

    content.push({
      type: "text",
      text: `Please OCR all content above. Output format: ${outputFormat}.`,
    });

    const response = await this.client.messages.create({
      model: this.model,
      max_tokens: 8192,
      system: SYSTEM_PROMPT,
      messages: [{ role: "user", content }],
    });

    const block = response.content[0];
    return block.type === "text" ? block.text : "";
  }
}

File: ocr-cli/src/ocr_cli/providers/anthropic.py
"""Anthropic Claude vision provider."""

import base64
from typing import Any

import anthropic

from ocr_cli.providers.base import BaseProvider
from ocr_cli.prompt import HANDWRITTEN_NOTES_PROMPT

SYSTEM_PROMPT = HANDWRITTEN_NOTES_PROMPT


class AnthropicProvider(BaseProvider):
    def __init__(self, api_key: str, model: str) -> None:
        self.client = anthropic.Anthropic(api_key=api_key)
        self.model = model

    def ocr(self, images: list[bytes], output_format: str = "markdown") -> str:
        content: list[Any] = []

        for i, img_bytes in enumerate(images):
            if len(images) > 1:
                content.append({"type": "text", "text": f"[Page {i + 1}]"})
            b64 = base64.standard_b64encode(img_bytes).decode("utf-8")
            content.append({
                "type": "image",
                "source": {
                    "type": "base64",
                    "media_type": "image/png",
                    "data": b64,
                },
            })

        content.append({
            "type": "text",
            "text": f"Please OCR all content above. Output format: {output_format}.",
        })

        response = self.client.messages.create(
            model=self.model,
            max_tokens=8192,
            system=SYSTEM_PROMPT,
            messages=[{"role": "user", "content": content}],
        )

        return response.content[0].text
